spec/ic.did

```diff
@@ -90,7 +90,7 @@ service ic : {
     settings : canister_settings
   }) -> ();
   install_code : (record {
-    mode : variant {install; reinstall; upgrade};
+    mode : variant {install; reinstall; upgrade; evict};
```

Comment:
@Dfinity-Bjoern @roman-kashitsyn wdyt?

Reply:
Ping? :)

Reply:
May also make sense whether the bug has been fixed in the meantime? @chenyan-dfinity

Reply:
Both extending `upgrade` to `upgrade : opt record` and adding a new variant tag in the function argument are non-breaking changes. This is always supported and doesn't rely on the extensible variant fix.

Reply:
> doesn't rely on the extensible variant fix

I believe the question was whether the "extensible variant" issue has been fixed in the meantime.

Reply:
@chenyan-dfinity The problem is not extending a function argument but the fact that these install code messages might need to fly between subnets and therefore we would also need to decode them on the receiving subnet. This is the problem with extending the variant tag and not that we'd break the interface of the management canister.

Reply:
I don't see the difference. At the Candid level, everything proposed here is fine, it wouldn't break any existing clients, including the receiving subnet who is running an old client. This is an allowed change even without applying the extensible variant fix, which means it's working in production without the need to bump to candid 0.9.

Reply:
Out of curiosity, I've run the following code:
```
use candid::{CandidType, Decode, Deserialize, Encode};
use serde::Serialize;

#[derive(CandidType, Deserialize, Serialize, Debug)]
pub enum CanisterInstallMode {
    #[serde(rename = "install")]
    Install = 1,
    #[serde(rename = "reinstall")]
    Reinstall = 2,
    #[serde(rename = "upgrade")]
    Upgrade = 3,
}

#[derive(CandidType, Deserialize, Serialize, Debug)]
pub enum CanisterInstallModeNew {
    #[serde(rename = "install")]
    Install = 1,
    #[serde(rename = "reinstall")]
    Reinstall = 2,
    #[serde(rename = "upgrade")]
    Upgrade = 3,
    #[serde(rename = "evict")]
    Evict = 4,
}

fn main() {
    let bytes = Encode!(&CanisterInstallMode::Install).unwrap();
    let mode_install = Decode!(&bytes, CanisterInstallModeNew).unwrap();
    let bytes = Encode!(&CanisterInstallMode::Reinstall).unwrap();
    let mode_reinstall = Decode!(&bytes, CanisterInstallModeNew).unwrap();
    let bytes = Encode!(&CanisterInstallMode::Upgrade).unwrap();
    let mode_upgrade = Decode!(&bytes, CanisterInstallModeNew).unwrap();
    println!(
        "install: {:?}; reinstall: {:?}; upgrade: {:?}",
        mode_install, mode_reinstall, mode_upgrade
    );
}
```
with the following `Cargo.toml`:
```
[package]
name = "subtyping"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
candid = "0.8"
ic-cdk = "0.7"
serde = "1"
```
and it prints
```
install: Install; reinstall: Reinstall; upgrade: Upgrade
```
as expected => the decoding seems to work fine.

Reply:
Yep, `ModeNew` can also be changed to 
```
#[derive(CandidType, Deserialize, Debug)]
pub struct Flag { skip: bool }

#[derive(CandidType, Deserialize, Debug)]
pub enum CanisterInstallModeNew {
    #[serde(rename = "install")]
    Install,
    #[serde(rename = "reinstall")]
    Reinstall,
    #[serde(rename = "upgrade")]
    Upgrade(Option<Flag>),
    #[serde(rename = "evict")]
    Evict,
}
```
and still decodes fine.

Reply:
@mraszyk You did not try to decode with the new variant "evict" using the old enum. That would not work. The problem is decoding an encoded blob containing the new variant while you're running the old code that doesn't contain it. I should've been more clear with my example of the two subnets communicating.

Here's the slack [thread](https://dfinity.slack.com/archives/C01DB8MQ5M1/p1675881598055839?thread_ts=1675868384.037619&cid=C01DB8MQ5M1) where I had attempted to add a new variant to `CanisterInstallMode` and we got an error because a subnet was sending an `install_code` with the new variant to another one that had the old version (without the new variant).

To confirm, use this modified main from your example:
```

Reply:
Continuing on @dsarlis message from above, we have two possible solutions:
1. Adding the new version of the `CanisterInstallMode` that would be the same as the old one while having the `Evict` mode, and the replica will always try to decode the newer version and if it fails, a replica will try to decode as the old version. The good thing about this approach is that code will be clean, and later we can remove the old version of the `CanisterInstallMode`. The bad thing is that this approach requires significant overhead because the `CanisterInstallMode` appears in a lot of the places in the replica and `NNS` canisters. Hence the plan will be, in the first release we should add a new version of `CanisterInstallMode`, in the second we should update the `NNS` canisters and start using `Evict` mode, and in the third release, we should remove the old `CanisterInstallMode`.
2. The other option is to add a flag to the `InstallCodeArgs` e.g. `skip_pre_upgrade`. The bad thing about this approach is that we will have the flag which exists for all the modes, but it applies only to one specific mode (`Upgrade`). The good thing is that this approach should be more straightforward, and easy to implement.
Wdyt @mraszyk @Dfinity-Bjoern ?

Reply:
> The bad thing about this approach is that we will have the flag which exists for all the modes, but it applies only to one specific mode (Upgrade).

How does this flag apply to all modes? It's only the property of the `upgrade` variant, so it only exists/applies to that mode.

Reply:
I see. In this case, adding `evict` needs to be done in two steps: 1) Upgrade `mode` type to `opt variant` from `variant`; 2) Use Candid 0.9, and add the new `evict` tag to `mode`.

Another more hacky but easier solution is @dragoljub-duric's option 2. Upgrade `install_code` to be of type `(record {...}, opt bool)`, while the last `opt bool` only applies to `Mode::Upgrade`.

Reply:
> The problem is decoding an encoded blob containing the new variant while you're running the old code that doesn't contain it.

@dsarlis This is a use case we never thought about. For canisters, we assume the upgrade is always done in one atomic step. Can we delay the client code upgrade until all subnets have caught up with the new replica? Otherwise, we essentially require both forward and backward compatibility, which is quite cumbersome to work with.

Reply:
> This is a use case we never thought about. For canisters, we assume the upgrade is always done in one atomic step.

Tbh, I think this is a miss on Candid's part. In a distributed setting, having an IDL that supports backward and forward compatibility is a must imo. We can't/don't control all clients or servers and when they update, so it's very useful when you get these properties on the IDL level.

> Otherwise, we essentially require both forward and backward compatibility, which is quite cumbersome to work with.

I would actually say quite the opposite, that not having a way to easily get forward compatibility with Candid is a big PITA. This is in theory a trivial change ("add a new variant to an enum") and we need to bend our backs to do it in a way that does not break clients in a guaranteed way.

> If this feature is only rolled out to one subnet, I think it's okay to error out for other subnets?

Erroring out might be ok in some circumstances but not in others. If there's anything that starts failing because an error was returned during decoding which was somehow not expected, then that would be bad. It's hard for us to vouch that all usages are properly handling errors returned, so I'm leaning towards an approach that gives us the safety even if it means some more work/extra steps that need to happen. Add on top of this that we have an awkward dependency of the NNS canisters on the types used by the replica's management canister (that we should eventually remove) and the whole thing becomes even more convoluted.

Reply:
> Add on top of this that we have an awkward dependency of the NNS canisters on the types used by the replica's management canister (that we should eventually remove) and the whole thing becomes even more convoluted.

Okay, that's the real problem. NNS is a client of the management canister. It should use the types generated from the deployed interface. Sharing types between service and client is making things unnecessarily difficult. In fact, I would say NNS canister should not be in the same repo as the replica to make testing easier.

In the real world, nobody upgrades the client before the new API is ready on the server side. It's expected for the old client to continue to work, while the new API is rolling out on the server side. But it's not expected for the server to respond to a future change that it is not aware of. Such system doesn't make sense mathematically.

Reply:
> Okay, that's the real problem. NNS is a client of the management canister. It should use the types generated from the deployed interface. Sharing types between service and client is making things unnecessarily difficult. In fact, I would say NNS canister should not be in the same repo as the replica to make testing easier.

Yes, indeed, that's a problem as I said and I've personally been pushing the NNS team to change this (as part of another cleanup work that I believe is super important, specifically to use the external Rust cdk instead of the internal one). As for being in a separate repo -- I have no skin in that game, therefore no strong opinion.

> In the real world, nobody upgrades the client before the new API is ready on the server side. It's expected for the old client to continue to work, while the new API is rolling out on the server side. But it's not expected for the server to respond to a future change that it is not aware of.

Mostly agree except for the fact that you might have "peer-to-peer" architectures instead of true client-server. The case of replicas talking to each other is one such case but there are more in the real world. In that case, there's no "server updates before client" because basically the same process can be client _and_ server depending on the context. This is why protobuf, thrift, etc also handle forward compatibility because in a distributed system the concepts of client and server are some times blended.

> Such system doesn't make sense mathematically.

It does engineeringly (it's an actual world, I googled it) :)

Reply:
I am ok with both, leaning towards adding the optional bool to upgrade, especially given that there's basically no difference in how we should handle the changes to avoid breakages.

Reply:
I'm also fine with both extending the enum and turning the upgrade variant into a variant with an attached record. I wouldn't extend the arguments of install_code.

What hasn't been explicitly discussed though is how we deal with None values of Option<CanisterInstallMode> that we can get in step 2 of Yan Chen's proposal because it seems we're stuck with Option<CanisterInstallMode> forever then, right?

I didn't get the sentence "the replica will always try to decode the newer version and if it fails, a replica will try to decode as the old version"; in particular, what is the exact new version and old version here.

Reply:
> when receiving they will first try to decode it as CanisterInstallMode_V2 and if it fails they will try to decode as CanisterInstallMode

in both cases a CanisterInstallMode can be decoded as CanisterInstallMode_V2 so decoding CanisterInstallMode_V2 can never fail => we don't need phase 3 and can already drop CanisterInstallMode in step 2

Reply:
I like the `skip_pre_upgrade` idea, but I'd make the field optional as well (just in case):

```
      upgrade : opt record {
        skip_pre_upgrade: opt bool;
      }
```

Reply:
So just to make sure we're all on the same page: the current suggestion is to essentially follow @ielashi's proposal above (https://github.com/dfinity/interface-spec/pull/135#discussion_r1212896298) but with @roman-kashitsyn's tweak making `skip_pre_upgrade` an `opt bool`? That seems fine by me.

Reply:
I've changed the type of the field `skip_pre_upgrade` from `bool` to `opt bool` and thus resolve this thread.